kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,271 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+                 private_event_group_ids.end(),
+                 std::back_inserter(eg_state->event_group_id_batch));
+      // We cannot return the full merge
+      // E.g.: a=[1,2,3,4,5] b=[6,7,8,9]
+      if (eg_state->event_group_id_batch.size() > kBatchSize) {
+        eg_state->event_group_id_batch.erase(eg_state->event_group_id_batch.begin() + kBatchSize,
+                                             eg_state->event_group_id_batch.end());
+      }
+      // find and update the offsets \
+      // We iterate over event_group_id_batch in the reverse order, and update the private/public offset if the event group id
+      // in event_group_id_batch is found in private_event_group_ids/public_event_group_ids respectively.\
+      // In the worst case, we need to read the entire event_group_id_batch (of size kBatchSize) to find the offset
+      bool has_pub_offset_updated = false;
+      bool has_pvt_offset_updated = false;
+      for (auto r_it = eg_state->event_group_id_batch.rbegin(); r_it != eg_state->event_group_id_batch.rend(); r_it++) {
+        if (has_pub_offset_updated && has_pvt_offset_updated) break;
+        if (!has_pub_offset_updated && std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it) !=
+                                           public_event_group_ids.end()) {
+          auto pub_it = std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it);
+          eg_state->public_offset = ++pub_it - public_event_group_ids.begin();
+          has_pub_offset_updated = true;
+          LOG_DEBUG(logger_,
+                    "Updated public_offset: " << eg_state->public_offset
+                                              << " public_event_group_ids size: " << public_event_group_ids.size());
+        }
+        if (!has_pvt_offset_updated &&
+            std::find(private_event_group_ids.begin(), private_event_group_ids.end(), *r_it) !=
+                private_event_group_ids.end()) {
+          auto pvt_it = std::find(private_event_group_ids.begin(), private_event_group_ids.end(), *r_it);
+          eg_state->private_offset = ++pvt_it - private_event_group_ids.begin();
+          has_pvt_offset_updated = true;
+          LOG_DEBUG(logger_,
+                    "Updated private_offset: " << eg_state->private_offset
+                                               << " private_event_group_ids size: " << private_event_group_ids.size());
+        }
+      }
+    }
+    // let's not return an invalid event group id
+    if (eg_state->event_group_id_batch.empty()) return std::nullopt;
+
+    // Reset the iterator so that we read from event_group_id_batch from the beginning
+    eg_state->it = eg_state->event_group_id_batch.begin();
+
+    ConcordAssert(eg_state->event_group_id_batch.size() <= kBatchSize);
+    LOG_DEBUG(logger_, "Updated event group id list batch size: " << eg_state->event_group_id_batch.size());
+  }
+  return *eg_state->it++;
```

Comment:
The if-block is huge. For readability, can we turn the outermost if-condition around and return `*eg_state->it++` if we didn't reach the end. Then all the code above has one level of indentation less.